File: docs/palindromes-in-peg/delayed_pal.py
"""Two overlapping dyadic stages for a directly clocked PAL experiment.

This accepts original words, but is still an indexed reference machine, not
an ordinary PEG. Rates below bound yielded indexed operations. Local-head
lowering and its own service bound must precede SCA/PEG compilation.
"""
from gs_events import PatternMatcher, palindrome_job


MATCH_RATE = 80
JOB_RATE = 1024


class Window:
  def __init__(self, data, start, end=None, reverse=False):
    self.data, self.start, self.end, self.reverse = data, start, end, reverse
    if start < 0 or end is not None and not start <= end <= len(data):
      raise ValueError("window must contain only arrived characters")
    if reverse and end is None:
      raise ValueError("a reversed window needs a fixed endpoint")

  def __len__(self):
    return (len(self.data) if self.end is None else self.end) - self.start

  def __getitem__(self, index):
    if not 0 <= index < len(self):
      raise IndexError(index)
    return self.data[self.end - 1 - index if self.reverse else self.start + index]


class Stage:
  def __init__(self, data, width):
    self.data, self.width = data, width
    self.matcher = PatternMatcher(Window(data, 0, width, True), Window(data, width))
    self.job = None
    self.job_index = None
    self.job_flags = None
    self.results = [None] * 4
    self.job_steps = self.match_steps = 0
    self.jobs_completed = 0

  def tick(self, now):
    width, half = self.width, self.width // 2
    if now > width and (now - width) % half == 0:
      batch = (now - width) // half - 1
      if batch < 4:
        if self.job is not None:
          raise AssertionError("offline palindrome job overran its release interval")
        self.job_index, self.job_flags = batch, []
        size = (batch + 1) * half
        self.job = palindrome_job(Window(self.data, width, width + size), batch * half, size)
    if self.job is not None:
      for _ in range(JOB_RATE):
        event = self.job.step()
        self.job_steps += 1
        if event is not None and event[0] == "flag":
          self.job_flags.append(event[1])
        if self.job.done:
          if len(self.job_flags) != half:
            raise AssertionError("incomplete palindrome flag block")
          self.results[self.job_index] = self.job_flags
          self.job = self.job_flags = None
          self.jobs_completed += 1
          break
    matched = False
    for _ in range(MATCH_RATE):
      if self.matcher.waiting:
        break
      endpoint = self.matcher.step()
      self.match_steps += 1
      if endpoint is not None:
        if endpoint + width != now:
          raise AssertionError("positive pattern match missed its arrival deadline")
        matched = True
    if now < 2 * width:
      return None
    batch = (now - 2 * width) // half
    flags = self.results[batch]
    if not flags:
      raise AssertionError("middle palindrome flags missed their deadline")
    middle = flags.pop()
    return matched and middle


class DelayedPal:
  """Feed one binary character; report that exact prefix's PAL membership."""
  def __init__(self):
    self.data, self.stages = [], []
    self.next_width = 2
    self.retired_job_steps = self.retired_match_steps = 0
    self.retired_jobs = 0

  def feed(self, char):
    if char not in ("a", "b"):
      raise ValueError("binary input required")
    self.data.append(char)
    now = len(self.data)
    while self.stages and now == 4 * self.stages[0].width:
      retired = self.stages.pop(0)
      self.retired_job_steps += retired.job_steps
      self.retired_match_steps += retired.match_steps
      self.retired_jobs += retired.jobs_completed
    if now == self.next_width:
      self.stages.append(Stage(self.data, self.next_width))
      self.next_width *= 2
    result = True if now < 2 else self.data[0] == char if now < 4 else None
    for stage in self.stages:
      answer = stage.tick(now)
      if answer is not None:
        if result is not None:
          raise AssertionError("overlapping active answer intervals")
        result = answer
    if len(self.stages) > 2 or result is None:
      raise AssertionError("dyadic stage coverage failed")
    return result

  def statistics(self):
    return dict(characters=len(self.data),
                job_steps=self.retired_job_steps + sum(s.job_steps for s in self.stages),
                match_steps=self.retired_match_steps + sum(s.match_steps for s in self.stages),
                completed_jobs=self.retired_jobs + sum(s.jobs_completed for s in self.stages),
                live_stages=len(self.stages))


def recognize(word):
  machine, answer = DelayedPal(), True
  for char in word:
    answer = machine.feed(char)
  return answer

File: scala/pal/src/main/scala/pal/DelayedPal.scala
package pal

import scala.collection.IndexedSeq
import scala.collection.immutable.VectorMap
import scala.collection.mutable

/** A live view of a slice of the arrived characters; a growing window has no
  * fixed end, a reversed window reads its fixed slice backwards.
  */
final class Window(val data: mutable.ArrayBuffer[Char], val start: Int, val end: Option[Int] = None,
                   val backwards: Boolean = false) extends IndexedSeq[Char] {
  if (start < 0 || end.exists(stop => !(start <= stop && stop <= data.length))) {
    throw new IllegalArgumentException("window must contain only arrived characters")
  }
  if (backwards && end.isEmpty) {
    throw new IllegalArgumentException("a reversed window needs a fixed endpoint")
  }

  def length: Int = end.getOrElse(data.length) - start

  def apply(index: Int): Char = {
    if (!(0 <= index && index < length)) {
      throw new IndexOutOfBoundsException(index.toString)
    }
    if (backwards) data(end.get - 1 - index) else data(start + index)
  }
}

/** Two overlapping dyadic stages for a directly clocked PAL experiment.
  *
  * This accepts original words, but is still an indexed reference machine, not
  * an ordinary PEG. Rates below bound yielded indexed operations. Local-head
  * lowering and its own service bound must precede SCA/PEG compilation.
  *
  * Port of `delayed_pal.py`; the schedule is explained in `DELAYED_PAL.md`.
  */
object DelayedPal {
  import GsEvents.{IndexedJob, PatternMatcher, palindromeJob}

  val MATCH_RATE: Int = 80
  val JOB_RATE: Int = 1024

  /** One dyadic stage of width `K`: an outer GS matcher for the first `K`
    * characters reversed, and four sequential offline middle-flag jobs.
    */
  final class Stage(val data: mutable.ArrayBuffer[Char], val width: Int) {
    val matcher: PatternMatcher[Char] = new PatternMatcher(new Window(data, 0, Some(width), backwards = true), new Window(data, width))
    var job: Option[IndexedJob[Option[Char], Unit]] = None
    var jobIndex: Int = -1
    var jobFlags: mutable.ArrayBuffer[Boolean] = mutable.ArrayBuffer.empty
    val results: Array[Option[mutable.ArrayBuffer[Boolean]]] = Array.fill(4)(None)
    var jobSteps: Int = 0
    var matchSteps: Int = 0
    var jobsCompleted: Int = 0

    private def releaseJob(now: Int): Unit = {
      val half = width / 2
      if (now > width && (now - width) % half == 0) {
        val batch = (now - width) / half - 1
        if (batch < 4) {
          if (job.isDefined) {
            throw new AssertionError("offline palindrome job overran its release interval")
          }
          jobIndex = batch
          jobFlags = mutable.ArrayBuffer.empty
          val size = (batch + 1) * half
          job = Some(palindromeJob(new Window(data, width, Some(width + size)), batch * half, size))
        }
      }
    }

    private def serviceJob(): Unit = {
      job.foreach { active =>
        var budget = JOB_RATE
        while (budget > 0 && job.isDefined) {
          budget -= 1
          val event = active.step()
          jobSteps += 1
          event match {
            case Some(IndexedOp.Flag(value)) => jobFlags += value
            case _ => ()
          }
          if (active.done) {
            if (jobFlags.length != width / 2) {
              throw new AssertionError("incomplete palindrome flag block")
            }
            results(jobIndex) = Some(jobFlags)
            job = None
            jobFlags = mutable.ArrayBuffer.empty
            jobsCompleted += 1
          }
        }
      }
    }

    private def serviceMatcher(now: Int): Boolean = {
      var matched = false
      var budget = MATCH_RATE
      while (budget > 0 && !matcher.waiting) {
        budget -= 1
        val endpoint = matcher.step()
        matchSteps += 1
        endpoint.foreach { end =>
          if (end + width != now) {
            throw new AssertionError("positive pattern match missed its arrival deadline")
          }
          matched = true
        }
      }
      matched
    }

    /** Serve one arrival; the answer for the current prefix while this stage is
      * responsible for it (`2K <= now < 4K`).
      */
    def tick(now: Int): Option[Boolean] = {
      releaseJob(now)
      serviceJob()
      val matched = serviceMatcher(now)
      if (now < 2 * width) {
        None
      } else {
        val batch = (now - 2 * width) / (width / 2)
        val flags = results(batch).filter(_.nonEmpty).getOrElse {
          throw new AssertionError("middle palindrome flags missed their deadline")
        }
        val middle = flags.remove(flags.length - 1)
        Some(matched && middle)
      }
    }
  }

  /** Feed one binary character; report that exact prefix's PAL membership. */
  final class DelayedPal {
    val data: mutable.ArrayBuffer[Char] = mutable.ArrayBuffer.empty
    val stages: mutable.ArrayBuffer[Stage] = mutable.ArrayBuffer.empty
    var nextWidth: Int = 2
    var retiredJobSteps: Int = 0
    var retiredMatchSteps: Int = 0
    var retiredJobs: Int = 0

    private def retireStages(now: Int): Unit = {
      while (stages.nonEmpty && now == 4 * stages.head.width) {
        val retired = stages.remove(0)
        retiredJobSteps += retired.jobSteps
        retiredMatchSteps += retired.matchSteps
        retiredJobs += retired.jobsCompleted
      }
    }

    def feed(char: Char): Boolean = {
      if (char != 'a' && char != 'b') {
        throw new IllegalArgumentException("binary input required")
      }
      data += char
      val now = data.length
      retireStages(now)
      if (now == nextWidth) {
        stages += new Stage(data, nextWidth)
        nextWidth *= 2
      }
      var result: Option[Boolean] = if (now < 2) Some(true) else if (now < 4) Some(data(0) == char) else None
      stages.foreach { stage =>
        stage.tick(now).foreach { answer =>
          if (result.isDefined) {
            throw new AssertionError("overlapping active answer intervals")
          }
          result = Some(answer)
        }
      }
      if (stages.length > 2 || result.isEmpty) {
        throw new AssertionError("dyadic stage coverage failed")
      }
      result.get
    }

    /** Python's insertion-ordered dict of counters. */
    def statistics(): VectorMap[String, Int] = {
      VectorMap(
        "characters" -> data.length,
        "job_steps" -> (retiredJobSteps + stages.map(_.jobSteps).sum),
        "match_steps" -> (retiredMatchSteps + stages.map(_.matchSteps).sum),
        "completed_jobs" -> (retiredJobs + stages.map(_.jobsCompleted).sum),
        "live_stages" -> stages.length
      )
    }
  }

  def recognize(word: String): Boolean = {
    val machine = new DelayedPal
    var answer = true
    word.foreach(char => answer = machine.feed(char))
    answer
  }
}
